cmd/tunneld/tunneld.go

```diff
@@ -138,3 +150,87 @@ func fatal(format string, a ...interface{}) {
 	fmt.Fprint(os.Stderr, "\n")
 	os.Exit(1)
 }
+
+func startAutocert(opts *options, server *tunnel.Server, logger log.Logger) {
+	cacheDir, err := cacheDir(opts.letsEncryptCacheDir)
+	if err != nil {
+		fatal("failed to set cache directory for certs ", err)
+	}
+	m := &autocert.Manager{
+		Prompt:     autocert.AcceptTOS,
+		HostPolicy: hostPolicy(server),
+		Cache:      autocert.DirCache(cacheDir),
+	}
+	if opts.httpAddr == ":80" {
+		opts.httpAddr = ""
+	}
+	if opts.httpsAddr == ":443" {
+		opts.httpsAddr = ""
+	}
+	// Allow binding to a specific host
+	if !validAddr(opts.httpAddr) || !validAddr(opts.httpsAddr) {
+		fatal("invalid  http(s) address, port not allowed with Let's Encrypt enabled")
+	}
+	httpAddr := net.JoinHostPort(opts.httpAddr, "80")
+	httpsAddr := net.JoinHostPort(opts.httpsAddr, "443")
+	s := &http.Server{
+		Addr:    httpsAddr,
+		Handler: server,
+	}
+	s.TLSConfig = &tls.Config{
+		GetCertificate: m.GetCertificate,
+	}
+	http2.ConfigureServer(s, nil)
+	logger.Log(
+		"level", 1,
+		"action", "start https with Let's Encrypt support",
+		"addr", "80,443",
+	)
+	go func() {
+		fatal("failed to start HTTP: %s", http.ListenAndServe(httpAddr, m.HTTPHandler(http.HandlerFunc(redirect))))
+	}()
+	go func() {
+		fatal("failed to start HTTPS: %s", s.ListenAndServeTLS("", ""))
+	}()
+}
+
+func hostPolicy(server *tunnel.Server) autocert.HostPolicy {
+	return func(_ context.Context, host string) error {
+		_, _, subscribed := server.Subscriber(host)
+		if !subscribed {
+			return fmt.Errorf("acme/autocert: host `%s` not subscribed", host)
+		}
+		return nil
+	}
+}
+
+// validAddr checks that addrs doesn't contains a port already
+func validAddr(addr string) bool {
```

Comment:
Using a regexp is not gonna work, `net.SplitHostPort` is your friend. User must be able to provide:

* "IP:443"
* "IP"
* ":443"

Similarly for HTTP.

Reply:
I did this approach based on your previous comments where if let's encrypt is enabled we shouldn't allow passing a port, so valid inputs will result in errors while splitting see. https://play.golang.org/p/flRDKtg13v4

Reply:
If someone uses port 443 explicitly it's still OK.

Based on the implementation we shall see if error contains "missing port in address"... if yes then fine (host only), if it returns without errors check if port is 443 (or 80).